rust/src/connection.rs

```diff
@@ -0,0 +1,56 @@
+use crate::config;
+use crate::wallet;
+use anyhow::Result;
+
+/// blocks the current thread, if you want this to be run asynchronously you need to spawn a thread
+/// from the outside.
+pub async fn keep_alive() -> Result<()> {
+    let mut connected = false;
+    loop {
+        let alive = check_alive().await?;
+
+        if !connected || !alive {
+            connected = connect().await?;
+        }
+
+        // looping here indefinitely to keep the connection with the maker alive.
+        tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+    }
+}
+
+async fn check_alive() -> Result<bool> {
+    tracing::trace!("Checking if maker is alive");
+    let client = reqwest::Client::builder()
+        .timeout(config::TCP_TIMEOUT)
+        .build()?;
+    let result = client
+        .get(config::maker_endpoint() + "/api/alive")
+        .send()
+        .await;
+
+    let alive = match result {
+        Ok(_) => true,
+        Err(err) => {
+            tracing::warn!("Maker is offline! {err:?}");
+            false
+        }
+    };
+
+    Ok(alive)
+}
+
+async fn connect() -> Result<bool> {
```

Comment:
🙃 Maybe a bit more idiomatic to return a `Result<()>` here.

It would simplify the code and above you could do:

```rust
connected = connect().await.is_ok();
```

---

As a side note, currently this function is infallible, so you don't need the `Result` in it's current form!

Reply:
💯 thanks for the feedback, I played around with it a bit, but opted against it to not risk introducing some regression.